File: Data-Structures-Algorithms-Databases/week-8/quick_sort.py
class Solution:
    def sortArray(self, nums: list[int]) -> list[int]:
        # using quick sort
        def quick_sort(nums, left=0, right=len(nums)-1):
            if left < right:
                pivot_index = partition(nums, left, right)
                # sort the smaller elements on the left side
                quick_sort(nums, left, pivot_index - 1)
                # sort the larger elements on the right side
                quick_sort(nums, pivot_index + 1, right)

        def partition(nums, left, right):
            pivot = nums[left]
            low = left + 1
            high = right
            while True:
                while low <= high and nums[high] >= pivot:
                    high -= 1
                while low <= high and nums[low] <= pivot:
                    low += 1
                if low <= high:
                    nums[low], nums[high] = nums[high], nums[low]
                else:
                    break
            nums[left], nums[high] = nums[high], nums[left]
            return high

        quick_sort(nums)
        return nums


if __name__ == "__main__":
    nums = [-4, 0, 7, 4, 9, -5, -1, 0, -7, -1]
    print("Original array:", nums)
    solution = Solution()
    solution.sortArray(nums)
    print("Sorted array:", nums)

File: Data-Structures-Algorithms-Databases/week-8/quick-sort.c
/*
To compile the code:
    // Linux/MacOS/Windows:
    gcc -o quick-sort quick-sort.c
    // Linux/MacOS:
    make quick-sort
To run the code:
    ./quick-sort
*/

#include <stdio.h>

void quicksort_run(int arr[], int left , int right);
int partition(int arr[], int left , int right);

void quicksort(int arr[], int len){
    quicksort_run(arr, 0, len - 1);
}

void quicksort_run(int arr[], int left , int right) {
    if(left < right){
        int pivotindex = partition(arr, left , right);
        quicksort_run(arr, left, pivotindex - 1);
        quicksort_run(arr, pivotindex + 1, right);
    }
}

int partition(int arr[], int left , int right) {
    int size = right - left + 1;
    int arr_b[size]; // Create temporary array with correct size
    // pivotindex = choosePivot(a, left , right)
    int pivotindex = left; // Choosing the leftmost element as the pivot
    int pivot = arr[left];
    int a_count = left;
    int b_count = 1;
    for ( int i = left; i <= right; i++) {
        if ( i == pivotindex )
            arr_b[0] = arr[i];
        else if (arr[i] < pivot || 
                (arr[i] == pivot && i < pivotindex))
            arr[a_count++] = arr[i];
        else
            arr_b[b_count++] = arr[i];
    }
    for (int i = 0; i < b_count; i++)
        arr[a_count++] = arr_b[i];
    return right - b_count + 1;
}

int main() {
    int arr[] = {4, 5, 2, 7, 8, 1, 3, 6};
    int length = sizeof(arr)/sizeof(arr[0]);
    printf("Length of the array: %d\n",length);
    quicksort(arr, length);
    for (int i = 0; i < length; i++)
        printf("%d ", arr[i]);
    return 0;
}